kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/BufferedTensor.kt

```diff
@@ -0,0 +1,86 @@
+package space.kscience.kmath.tensors.core
+
+import space.kscience.kmath.nd.MutableBufferND
+import space.kscience.kmath.structures.*
+import space.kscience.kmath.tensors.api.TensorStructure
+import space.kscience.kmath.tensors.core.algebras.TensorLinearStructure
+
+
+public open class BufferedTensor<T>(
+    override val shape: IntArray,
+    internal val buffer: MutableBuffer<T>,
+    internal val bufferStart: Int
+) : TensorStructure<T> {
+    public val linearStructure: TensorLinearStructure
+        get() = TensorLinearStructure(shape)
+
+    public val numElements: Int
+        get() = linearStructure.size
+
+    override fun get(index: IntArray): T = buffer[bufferStart + linearStructure.offset(index)]
+
+    override fun set(index: IntArray, value: T) {
+        buffer[bufferStart + linearStructure.offset(index)] = value
+    }
+
+    override fun elements(): Sequence<Pair<IntArray, T>> = linearStructure.indices().map {
+        it to this[it]
+    }
+
+    override fun equals(other: Any?): Boolean = false
+
+    override fun hashCode(): Int = 0
+
+}
+
+public class IntTensor internal constructor(
+    shape: IntArray,
+    buffer: IntArray,
+    offset: Int = 0
+) : BufferedTensor<Int>(shape, IntBuffer(buffer), offset) {
+    internal constructor(bufferedTensor: BufferedTensor<Int>) :
```

Comment:
Should be removed in favor of functions below. Secondary constructors are discouraged.